src/modrover/rover.py

```diff
@@ -0,0 +1,118 @@
+from typing import Callable, Optional, Union
+
+import pandas as pd
+
+from .globals import get_rmse
+from .learner import Learner, LearnerID
+from .strategies.base import RoverStrategy
+from .strategies import get_strategy
+
+
+class Rover:
+
+    def __init__(
+        self,
+        model_type: str,
+        y: str,
+        cov_fixed: dict[str, list[str]],
+        cov_explore: dict[str, list[str]],
+        extra_param_specs: Optional[dict[str, dict]] = None,
+        offset: str = "offset",
+        weights: str = "weights",
+        holdout_cols: Optional[list[str]] = None,
+        model_eval_metric: Callable = get_rmse,
+    ) -> None:
+        # parse extra_param_specs
+        if extra_param_specs is None:
+            extra_param_specs = {}
+
+        self.model_type = model_type
+        self.y = y
+        self.cov_fixed = cov_fixed
+        self.cov_explore = cov_explore
+        self.extra_param_specs = extra_param_specs
+        self.offset = offset
+        self.weights = weights
+        self.holdout_cols = holdout_cols
+        self.model_eval_metric = model_eval_metric
+
+        self.learners: dict[LearnerID, Learner] = {}
+
+        # TODO: validate the inputs
+        # ...
+
+    @property
+    def performances(self):
+        """Convenience property to select only the performance of each learner."""
+        return {lid: learner.performance for lid, learner in self.learners.items()}
+
+    def get_learner(self, learner_id: LearnerID) -> Learner:
+
+        # See if we've already initialized one
+        if learner_id in self.learners:
+            return self.learners[learner_id]
+
+        all_covariates = list(self.cov_explore.values())[0]
+        param_specs = {}
+        for param_name, covs in self.cov_fixed.items():
+            variables = covs.copy()
+            if param_name in self.cov_explore:
+                variables.extend([
+                    # Ignore the always-present 0 index. Results in duplicate column names.
+                    all_covariates[i - 1] for i in learner_id.cov_ids[1:]
```

Comment:
Commenting that this line is responsible for the SingularDesignMatrix warnings I discussed yesterday. The 0-th index is kind of a dummy variable so we shouldn't be including it, including it gives us duplicated `Variable` objects attached to the resultant model

Reply:
Thanks David! We should discuss more this week.